### PR TITLE
Add true retriever lock

### PR DESCRIPTION
**Pull request by pengweiqhca**
Local lock: MemoryLock
Distributed Lock: CSRedisLock, MemcachedLock, RedisLock

### CODE REVIEW

**Comment by pengweiqhca** [comment]
Interceptor_Should_Recognize_Subclass_Of_EasyCachingAble_Attribute test fail.

File: src/EasyCaching.CSRedis/DistributedLock/CSRedisLockProvider.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            (long)_database.Eval(@"if redis.call('GET', KEYS[1]) == ARGV[1] then
+    return redis.call('DEL', KEYS[1]);
+end
+return -1;", $"{_name}/{key}", value) >= 0;
+
+        public async Task<bool> DeleteAsync(string key, byte[] value) =>
+            (long)await _database.EvalAsync(@"if redis.call('GET', KEYS[1]) == ARGV[1] then
+    return redis.call('DEL', KEYS[1]);
+end
+return -1;", $"{_name}/{key}", value) >= 0;
```
Comment (catcherwong) [inline src/EasyCaching.CSRedis/DistributedLock/CSRedisLockProvider.cs:38]: What about move the lua script to a constant?

**Review by catcherwong — APPROVED** [review]
LGTM, Thanks!

**Comment by catcherwong** [comment]
> Interceptor_Should_Recognize_Subclass_Of_EasyCachingAble_Attribute

This unit test is a bit unstable. I'll fix it later.

**Comment by catcherwong** [comment]
Please change the target branch from `master`  to `dev`.

File: src/EasyCaching.CSRedis/DistributedLock/CSRedisLockProvider.cs (hunk 1 above)
Comment (pengweiqhca) [reply src/EasyCaching.CSRedis/DistributedLock/CSRedisLockProvider.cs:38]: https://redis.io/topics/distlock

File: test/EasyCaching.UnitTests/DistributedLock/MemoryLockTest.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                            await Task.Delay(random.Next(0, 10));
+
+                            await @lock.ReleaseAsync().ConfigureAwait(false);
+                        }
+                        catch (DistributedLockException)
+                        {
+                        }
+                    }
+                })));
+        }
```
Comment (LabelZhou) [inline test/EasyCaching.UnitTests/DistributedLock/MemoryLockTest.cs:48]: What is the purpose of this test?

**Review by LabelZhou — APPROVED** [review]
LGTM

File: src/EasyCaching.Core/EasyCachingAbstractProvider.cs [hunk 3]
```diff
@@ -166,8 +181,31 @@ public CacheValue<T> Get<T>(string cacheKey, Func<T> dataRetriever, TimeSpan exp
...
+                {
+                    if (!@lock.Lock(_options.SleepMs)) throw new TimeoutException();
+
+                    value = BaseGet<T>(cacheKey);
+                    if (value.HasValue) return value;
+
+                    var item = dataRetriever();
+                    if (item != null || _options.CacheNulls)
+                    {
+                        BaseSet(cacheKey, item, expiration);
```
[the remark this answers is not in the document]
Comment (pengweiqhca) [reply src/EasyCaching.Core/EasyCachingAbstractProvider.cs:200]: BaseGet have no diagnostic log

File: src/EasyCaching.CSRedis/Configurations/CSRedisLockExtension.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        /// </summary>
+        /// <param name="name">Name.</param>
+        public CSRedisLockExtension(string name) => this._name = name;
+
+        /// <summary>
+        /// Adds the services.
+        /// </summary>
+        /// <param name="services">Services.</param>
+        public void AddServices(IServiceCollection services) =>
+            services.Replace(ServiceDescriptor.Singleton<IDistributedLockFactory, CSRedisLockFactory>(x =>
```
[the remark this answers is not in the document]
Comment (pengweiqhca) [reply src/EasyCaching.CSRedis/Configurations/CSRedisLockExtension.cs]: I fix this issure

File: test/EasyCaching.UnitTests/DistributedLock/MemoryLockTest.cs (hunk 2 above)
Comment (pengweiqhca) [reply test/EasyCaching.UnitTests/DistributedLock/MemoryLockTest.cs:48]: Except repeat lock, all case should be success.

**Pull request merged**